User signup/login form and db model validations, update dependencies

jeffgore00 opened the pull request:
This addresses subtasks 1, 2, and 4 in #162 .

The most significant change in this commit is integrating [Redux-Form](https://redux-form.com/6.0.0-alpha.4/docs/gettingstarted.md/) which adds a `form` property to state. Using its `Field` component (which itself can render any component we want, such as Semantic's `Input`) allows us to pass in an array of validating functions:

```js
import { Field, reduxForm } from 'redux-form';
import {
  isFilledOut,
  isEmail,
} from '../utils/formValidations';

<Field
  component={FieldWithWarning}
  name="email"
  type="text"
  placeholder="Email"
  validate={[isFilledOut, isEmail]} // <-----THIS
/>
```
In other words, we now have validation on the front end that the user is entering a valid email address, as well as a strong password. Password strength (and hints for better passwords) comes from the `zxcvbn` library.

I also updated the database model to add an `isEmail` validation to the the `email` field in the `User` model.

Finally, I updated the `mocha` version in the `package.json` because one of its dependencies, `growl`, had a security vulnerability in one of its earlier versions (per GitHub).

jeffgore00 commented:
Sorry for the delay, guys:

Addressing your points first, Don:

> When entering the wrong password the error message reads "Data must be a string or buffer". Can we change that to something more informative to the user?

I actually don't see that error when I enter the wrong password on a login. I see "Wrong username and/or password." Can you let me know exactly which conditions lead to this error? I can't seem to recreate it.

> `Do AuthForm and SignupOrLogin form need to be separate components? If so, maybe we should give them better names? Like 'AuthForm' and 'AuthFormFields' or something.`

I think they should remain separate components because each has substantial functionality and `AuthForm` would be too bloated (IMO) if it absorbed `SignupOrLogin`. Renaming is a good idea though, I renamed it `AuthFormLocal` since it deals specifically with handling local authorization.

> Would also be good to have the "Please use a stronger password" message be more informative (stronger how? longer? special chars? etc.) but that can be another pull request.

Sometimes you actually will get a specific warning. Try entering 'aaa' or 'jones' as a password and you'll see warnings about using repeating letters or names. But  a warning is not the same thing as a suggestion and wouldn't you know, `zxcvbn` supplies a `suggestions` array as well. Just like `warning`, it isn't guaranteed to be populated, but a lot of the times it is. So I've implemented the use of `suggestions` in my latest commit to address this concern. 

Anu:

> The password strength requirement is kind of ridiculous if you try out a bunch of combinations. My "strong" password on majority of websites still causes this to complain. Is there a way to tone this down using `zxcvbn`?

Anu, check out the `zxcvbn` [README](https://github.com/dropbox/zxcvbn) for their position on what constitutes a "strong" vs "weak" password: "Consider using zxcvbn as an algorithmic alternative to password composition policy — it is more secure, flexible, and usable when sites require a minimal complexity score in place of annoying rules like "passwords must contain three of {lower, upper, numbers, symbols}".... [Those] policies often fail both ways, allowing weak passwords (P@ssword1) and disallowing strong passwords."

What it all boils down to is that `zxcvbn` gives a password a score from 0 to 4, unsafest to safest. I dialed the minimum score down from 3 to 2, see if that helps. Here is what the password validation function looks like now:

```js
export const isStrongPassword = value => {
  const verdict = zxcvbn(value);
  const { feedback } = verdict;
  return verdict.score < 2 ? buildPasswordWarning(feedback) : undefined;
};
```
Note that returning `undefined` if the field data is valid is per [the specs for a Redux form validation function](https://redux-form.com/6.6.2/examples/fieldlevelvalidation/).

dleistman commented:
Cool. Yea Anu I'd definitely suggest reading the `zxcvbn` README, it's pretty interesting. I didn't realize that it was analyzing the complexity algorithmically rather than just checking off certain boxes (ie length, special chars, etc.). Honestly I'd say it's almost a bit of overkill since we don't really have sensitive data on our platform, but it seems fun and this is an educational project so whatever.

Anyway it looks much better now with the suggestions. It would maybe be cool to implement a password strength progress bar using the score it generates. If you guys are down with that, I'll make a ticket for it.

As for the "Data must be a string or buffer" error, this is actually an interesting edgecase, where the backend is throwing an error if you try to log in with an email address that you used to sign up for the site through Google OAuth, but in the regular login field. The error is being thrown on the backend, so I guess we can probably fix it with a try/catch block that then sends back a message along the lines of "Please log in with Google above".

Here's the stack trace from the server:

```js
TypeError: Data must be a string or a buffer
    at Hash.update (crypto.js:74:16)
    at Function.User.encryptPassword (/Users/Don/dev/fullstack/projects/readMe/repo/web-app/server/db/models/User.js:55:6)
    at model.User.correctPassword (/Users/Don/dev/fullstack/projects/readMe/repo/web-app/server/db/models/User.js:41:15)
    at User.findOne.then.user (/Users/Don/dev/fullstack/projects/readMe/repo/web-app/server/auth/index.js:11:24)
    at tryCatcher (/Users/Don/dev/fullstack/projects/readMe/repo/web-app/node_modules/bluebird/js/release/util.js:16:23)
    at Promise._settlePromiseFromHandler (/Users/Don/dev/fullstack/projects/readMe/repo/web-app/node_modules/bluebird/js/release/promise.js:512:31)
    at Promise._settlePromise (/Users/Don/dev/fullstack/projects/readMe/repo/web-app/node_modules/bluebird/js/release/promise.js:569:18)
    at Promise._settlePromise0 (/Users/Don/dev/fullstack/projects/readMe/repo/web-app/node_modules/bluebird/js/release/promise.js:614:10)
    at Promise._settlePromises (/Users/Don/dev/fullstack/projects/readMe/repo/web-app/node_modules/bluebird/js/release/promise.js:693:18)
    at Async._drainQueue (/Users/Don/dev/fullstack/projects/readMe/repo/web-app/node_modules/bluebird/js/release/async.js:133:16)
    at Async._drainQueues (/Users/Don/dev/fullstack/projects/readMe/repo/web-app/node_modules/bluebird/js/release/async.js:143:10)
    at Immediate.Async.drainQueues (/Users/Don/dev/fullstack/projects/readMe/repo/web-app/node_modules/bluebird/js/release/async.js:17:14)
    at runCallback (timers.js:666:20)
    at tryOnImmediate (timers.js:639:5)
    at processImmediate [as _immediateCallback] (timers.js:611:5)
```

jeffgore00 commented:
OK cool - I went ahead and implemented the "Please log in with Google above" message if a local login has a `googleId`.

I also:
- Modified the `AuthFormLocal` component so that any 500 server error displays a generic error message rather than the potentially cryptic message from the server (such as the one you encountered about the string/buffer).
- Removed the "Login with Twitter" button since we don't have Twitter OAuth.
- Modified `AuthForm` so that it is a class component with a `componentWillUnmount` hook to clear any login errors from state. I found that any login error message, which is tied to `user.error` on state, would persist after closing and re-opening a login or signup form.

I'm in support of password strength bar but I would prefer that live in a separate PR and that we try to merge this one.
